### ISSUE TITLE
[bug] Conanfiles loaded by path fail when named options override starred ones

### ISSUE THREAD

**Issue by rddesmond** [conan-io/conan#8830]
Thank you for the great work on conan.  When working with on developing recipes I found an issue with how profiles are used for requests by path.  I was using the `source(...)` function in the python api, but the CLI shown below has the same behavior.

When loading a conanfile,  options are processed differently if you take a path vs a reference.  Specifically, when you load by path, profile entries that are named (e.g. `ImportantRecipe:option=False`) will be ignored. The steps uses `conan info`, but I imagine any API call that can take a path would be susceptible.

For the error outcome below, it seems like having a stared property paired with the named one is a requirement.

### Environment Details (include every applicable attribute)
  * Operating System+version: CentOS Linux 7 (Core)
  * Conan version: Conan version 1.35.2 (also tested with the same behavior in 1.20 and 1.10)
  * Python version: Python 3.6.8

### Steps to reproduce (Include if Applicable)

1. Create a simple conanfile which has a some required option
    ```python
    class ImportantRecipe(ConanFile):
        name = "ImportantRecipe"
        version = '1.0'
        user = 'user'
        channel = 'channel'
        options = {"shared": [True]}
        default_options = {"shared": True}
    ```
2. Create a default profile that includes both a starred default and a package-specific variant.  (The order doesn't matter)
    ```
    [options]
    *:shared=False
    ImportantRecipe:shared=True
    ```
3. Export the recipe (this isn't required for the failure, but we'll use it to verify the reference version of commands work)
    ```
    conan export /path/to/importantrecipe/conanfile.py
    ```
4. Show that `conan info` for the reference works as expected
    ```
    conan info ImportantRecipe/1.0@user/channel
    ```
5. Test conan info for the file path. **_This step produces the unexpected behavior._**
    ```
    conan info /path/to/importantrecipe/conanfile.py
    ```
6. Update the default profile to include just an invalid named option.
    ```
    [options]
    ImportantRecipe:shared=False
    ```
7. Show that `conan info` for the reference fails as expected
    ```
    conan info ImportantRecipe/1.0@user/channel
    ```
8. Show that `conan info` for the file path fails as expected
    ```
    conan info /path/to/importantrecipe/conanfile.py
    ```

### Logs (Executed commands with output) (Include/Attach if Applicable)

Step 4 works as expected, and steps 7/8 fail as expected. Step 5 unexpectedly fails with:
```
ERROR: /path/to/importantrecipe/conanfile.py: 'False' is not a valid 'options.shared' value.
Possible values are ['True']
```

**Comment by memsharded** [comment]
Hi @rddesmond 

I am trying to reproduce this, but this only shows if the options is hardwired to only accept ``True``, right? Because the following test passes:

```python
def test_path_vs_ref_option():
    # https://github.com/conan-io/conan/issues/8830
    client = TestClient()
    conanfile = GenConanfile("hello", "0.1").with_option("shared", [True, False]).\
        with_default_option("shared", True)
    client.save({"conanfile.py": conanfile,
                 "profile": textwrap.dedent("""
                    [options]
                    *:shared=False
                    hello:shared=True
                    """)})
    client.run("export .")
    client.run("info hello/0.1@ -pr=profile")
    client.run("info. -pr=profile")
```

And will finally assign the correct shared=True value. It is not failing because it assigns the wrong value, but because as it matches the ``*`` pattern, it tries to assign the ``False`` value, and it fails because of incorrect value.

Why are you defining such option with only 1 value ``options = {"shared": [True]}``? That is not a real option if it doesn't allow more than 1 value?

**Comment by rddesmond** [comment]
Thank you for looking at it.  That's a useful check, and your look makes sense.  Perhaps the failure is because the option possibilities are checked before the `Name` is known, but it's good to know that it does the right thing if it gets past that.

1. Start loading the file
2. Initially determine and check that the options are satisfactory (without knowing its `Name` yet)
3. Finish loading the file, get the `Name` and requirements
4. Redetermine the options and check that they are satisfactory, now that you can look by Name

In that particular example, we have a standard "shared" option.  In the absence of that option, I'd assume that you'll get no additional libs that will have to be shipped with the executable (perhaps it's header-only or static).  This is a way of conveying that a recipe cannot be used to make lib-free binaries if you try by setting `*:shared=False`.  (You could have both options and a `verify()`, but that doesn't seem idiomatic.)  That said, I think that this specific example was added in anticipation of future static capability (and just removing the non-option works for us for now).

I could also imagine a more-than-binary option where most recipes accept (`A`, `B`, `C`), but some recipe only accepts (`A`, `B`).  If you want all `C` except that one, it would fail in the same way.

```python
def test_path_vs_ref_option():
    # https://github.com/conan-io/conan/issues/8830
    client = TestClient()
    conanfile = GenConanfile("hello", "0.1").with_option("opt", ['A', 'B']).\
        with_default_option("opt", 'A')
    client.save({"conanfile.py": conanfile,
                 "profile": textwrap.dedent("""
                    [options]
                    *:opt=C
                    hello:opt=B
                    """)})
    client.run("export .")
    client.run("info hello/0.1@ -pr=profile")
    client.run("info. -pr=profile")
```

**Comment by memsharded** [comment]
> Thank you for looking at it. That's a useful check, and your look makes sense. Perhaps the failure is because the option possibilities are checked before the Name is known, but it's good to know that it does the right thing if it gets past that.

Yes, the problem is that all matches are applied, first the pattern based ones, like ``*:shared`` then the specific one ``name:shared``, in this specific order. This is the necessary mechanism to have the more specific to have priority over the general one, but the general one should still apply if the other one was not matching. But that cannot be known in advance, so it is necessary to apply them sequentially. And ``*:shared=False`` is finding an invalid value there and raises.

>  (You could have both options and a verify(), but that doesn't seem idiomatic.) T

You mean the ``validate()`` method, right? Actually we are going to use it as the idiomatic one to restrict settings values, instead of the declaration. The ``validate()`` method has some advantages as it is delayed, so you can actually compute a graph, and print it, showing clearly which nodes in the graph are the invalid ones. Easier to debug for large graphs. I wouldn't say it is a bad approach for saying that your package cannot generate a static library.

**Comment by rddesmond** [comment]
I think the check being applied at the same time as `validate()` is a good idea, but think it's not intuitive that you need to specify all of the possible options for this very specific case (loading a conanfile by path), even where the source doesn't support it.

```python
class ImportantRecipe(ConanFile):
    name = "ImportantRecipe"
    version = '1.0'
    options = {"blas": ["A", "B", "C"]}
    default_options = {"blas": "A"}
    
    def validate(self):
       if self.options.blas == "C":
            raise ConanInvalidConfiguration("Blas C supported")
```

In the "shared" example (which I had), it's pretty easy to have the whole realm of possibilities, but what if it was something where the list was more freeform?  For instance, if you wanted something more general, like a blas implementation. If you wanted most recipes to be "C", but something didn't support that, so "A" was the fallback for it.  It seems like a valid use of the options file to be, I think?

```properties
[options]
*:blas=C
ImportantRecipe:blas=A"\
```

**Comment by memsharded** [comment]
I think this case is related to the semantics of wildcard option definition like ``*:blas=C``. Options are defined per package, but if there are some semantics for some known options that are used across different packages, then the definition of such option possible values should be the same, even if some packages do not implement later support for a given option (and might raise in ``validate()``). Maybe we should consider definition of class attributes a pure definition of the space, not the definition of the recipe support, in the same way as having ``settings = "os", "compiler", "arch"`` doesn't mean that this recipe will build for any os, compiler or architecture out there for any possible setting value.

The only possible alternative to this, would be to change the way profile options are applied, and use ordered definition priority. In that case you could put first the ``ImportantRecipe:blas=A`` one, and as it matches, it will be applied. The problem with this approach is that it is not possible to change it now for Conan 1.x, as it is breaking and will probably affect many users. If we decided to go this way, it should wait until Conan 2.0.

**Comment by rddesmond** [comment]
I don't think that I would change the priority that things are applied with, I think the "most specific" ordering that currently exists makes a lot of sense.  I also don't think it'd fix this, because the initial check is happening before the Name is known so it'd still match the wildcard in the first pass.

I agree that we are verging on a case where settings (or settings schematics) are more appropriate than options, but I think that this particular error is confusing, particularly when loading the same file by reference works as expected.  I think it would make sense to defer checking the options against its list until the time when `validate()` would be called, but looking at how it's implemented I haven't seen an intuitive way to do that.

**Comment by AbrilRBS** [comment]
Closing as Conan 1-only, for which support is no longer planned.

Note that Conan 2 has greatly improved option handling and issues like this should not show up anymore
